packages/cloud-cognitive/src/components/Datagrid/Datagrid/DatagridToolbar.js

```diff
@@ -1,29 +1,171 @@
-/*
- * Licensed Materials - Property of IBM
- * 5724-Q36
- * (c) Copyright IBM Corp. 2020
- * US Government Users Restricted Rights - Use, duplication or disclosure
- * restricted by GSA ADP Schedule Contract with IBM Corp.
+/**
+ * Copyright IBM Corp. 2022, 2022
+ *
+ * This source code is licensed under the Apache-2.0 license found in the
+ * LICENSE file in the root directory of this source tree.
  */
-import React from 'react';
-import { DataTable } from 'carbon-components-react';
-import { pkg } from '../../../settings';
+
+import React, { useEffect, useState } from 'react';
+import { Add16, OverflowMenuVertical16 } from '@carbon/icons-react';
+import {
+  DataTable,
+  TableBatchActions,
+  TableBatchAction,
+} from 'carbon-components-react';
+import { useResizeDetector } from 'react-resize-detector';
+import { ButtonMenu, ButtonMenuItem } from '../../ButtonMenu';
+import { pkg, carbon } from '../../../settings';
+import cx from 'classnames';
 
 const blockClass = `${pkg.prefix}--datagrid`;
 
 const { TableToolbar } = DataTable;
 
+const DatagridBatchActionsToolbar = (datagridState, width, ref) => {
+  const [displayAllInMenu, setDisplayAllInMenu] = useState(false);
+  const [initialListWidth, setInitialListWidth] = useState(null);
+  const [receivedInitialWidth, setReceivedInitialWidth] = useState(false);
+  const { selectedFlatRows, toggleAllRowsSelected, toolbarBatchActions } =
+    datagridState;
+  const totalSelected = selectedFlatRows && selectedFlatRows.length;
+
+  // Get initial width of batch actions container,
+  // used to measure when all items are put inside
+  // the ButtonMenu
+  useEffect(() => {
+    if (totalSelected === 1 && !receivedInitialWidth) {
+      const batchActionListWidth = ref.current.querySelector(
+        `.${carbon.prefix}--action-list`
+      ).offsetWidth;
+      setInitialListWidth(batchActionListWidth);
+      setReceivedInitialWidth(true);
+    }
+  }, [totalSelected, receivedInitialWidth, ref]);
+
+  useEffect(() => {
+    const summaryWidth = ref.current.querySelector(
+      `.${carbon.prefix}--batch-summary`
+    ).offsetWidth;
+    if (width < summaryWidth + initialListWidth + 32) {
+      setDisplayAllInMenu(true);
+    } else {
+      setDisplayAllInMenu(false);
+    }
+  }, [width, ref, initialListWidth]);
+
+  // Render batch actions in ButtonMenu
+  const renderBatchActionOverflow = () => {
+    const minWidthBeforeOverflowIcon = 380;
+    // Do not render ButtonMenu when there are 3 or less items
+    // and if there is enough available space to render all the items
+    if (toolbarBatchActions?.length <= 3 && !displayAllInMenu) {
+      return null;
+    }
+    return (
+      <ButtonMenu
+        label={width > minWidthBeforeOverflowIcon ? 'More' : null}
+        renderIcon={
+          width > minWidthBeforeOverflowIcon ? Add16 : OverflowMenuVertical16
```

Comment:
From the guidelines, it looks like the more icon should use a chevron instead, which also means we may need it to be directional?

Reply:
Unfortunately this (toggling icon based on open state) isn't possible with the `ButtonMenu` component _yet_ because of some underlying issues with the OverflowMenu from Carbon, but we do have an issue open https://github.com/carbon-design-system/ibm-cloud-cognitive/issues/2076 and also another with [Carbon](https://github.com/carbon-design-system/carbon/issues/11781). This is what @qbi11y has been working on